x/stake/keeper_test.go

```diff
@@ -259,36 +266,178 @@ func TestGetValidators(t *testing.T) {
+	validators := genValidators(candidates)
+
 	// test from nothing to something
-	// test from something to nothing
+	acc := keeper.getAccUpdateValidators(ctx)
+	assert.Equal(t, 0, len(acc))
+	keeper.setCandidate(ctx, candidates[0])
+	keeper.setCandidate(ctx, candidates[1])
+	_ = keeper.GetValidators(ctx) // to init recent validator set
+	acc = keeper.getAccUpdateValidators(ctx)
+	validatorsEqual(t, validators, acc)
+
 	// test identical
+	keeper.setCandidate(ctx, candidates[0])
+	keeper.setCandidate(ctx, candidates[1])
+	acc = keeper.getAccUpdateValidators(ctx)
+	validatorsEqual(t, validators, acc)
+
+	acc = keeper.getAccUpdateValidators(ctx)
+
+	// test from something to nothing
+	keeper.removeCandidate(ctx, candidates[0].Address)
+	keeper.removeCandidate(ctx, candidates[1].Address)
+	acc = keeper.getAccUpdateValidators(ctx)
+	assert.Equal(t, 2, len(acc))
+	assert.Equal(t, validators[0].Address, acc[0].Address)
+	assert.Equal(t, int64(0), acc[0].VotingPower.Evaluate())
+	assert.Equal(t, validators[1].Address, acc[1].Address)
+	assert.Equal(t, int64(0), acc[1].VotingPower.Evaluate())
+
 	// test single value change
+	candidates[0].Assets = sdk.NewRat(600)
+	validators = genValidators(candidates)
+	keeper.setCandidate(ctx, candidates[0])
+	keeper.setCandidate(ctx, candidates[1])
+	acc = keeper.getAccUpdateValidators(ctx)
+	validatorsEqual(t, validators, acc)
+
 	// test multiple value change
+	candidates[0].Assets = sdk.NewRat(200)
+	candidates[1].Assets = sdk.NewRat(0)
+	validators = genValidators(candidates)
+	keeper.setCandidate(ctx, candidates[0])
+	keeper.setCandidate(ctx, candidates[1])
+	acc = keeper.getAccUpdateValidators(ctx)
+	validatorsEqual(t, validators, acc)
+
 	// test validator added at the beginning
-	// test validator added in the middle
+	candidates = append([]Candidate{candidatesIn[0]}, candidates...)
+	validators = genValidators(candidates)
+	keeper.setCandidate(ctx, candidates[0])
+	keeper.setCandidate(ctx, candidates[1])
+	keeper.setCandidate(ctx, candidates[2])
+	acc = keeper.getAccUpdateValidators(ctx)
+	validatorsEqual(t, validators, acc)
+
+	// test validator added at the middle
+	candidates = []Candidate{candidates[0], candidates[1], candidatesIn[2], candidates[2]}
+	validators = genValidators(candidates)
+	keeper.setCandidate(ctx, candidates[0])
+	keeper.setCandidate(ctx, candidates[1])
+	keeper.setCandidate(ctx, candidates[2])
+	keeper.setCandidate(ctx, candidates[3])
+	acc = keeper.getAccUpdateValidators(ctx)
+	validatorsEqual(t, validators, acc)
+
 	// test validator added at the end
-	// test multiple validators removed
+	candidates = append(candidates, candidatesIn[4])
+	validators = genValidators(candidates)
+	keeper.setCandidate(ctx, candidates[0])
+	keeper.setCandidate(ctx, candidates[1])
+	keeper.setCandidate(ctx, candidates[2])
+	keeper.setCandidate(ctx, candidates[3])
+	keeper.setCandidate(ctx, candidates[4])
+	acc = keeper.getAccUpdateValidators(ctx)
+	validatorsEqual(t, validators, acc)
```

Comment:
Sorry I broke this test >:)  - lol but yeah again we need to reconfigure what we are even testing here!